src/lib/core/datetime.c

```diff
@@ -293,3 +293,105 @@ interval_to_string(const struct datetime_interval *ival, char *buf, ssize_t len)
 	}
 	return sz;
 }
+
+#define CHECK_RANGE(v, from, to) \
+	do { \
+		if ((v) < (from)) \
+			return -1; \
+		else if ((v) > (to)) \
+			return +1; \
+	} while (0)
+
+#define CHECK_DT(dt) \
+	CHECK_RANGE(dt, INT_MIN, INT_MAX)
+
+/**
+ * Normalize seconds and nanoseconds:
+ * - make sure that nanoseconds part is positive
+ * - make sure it's not exceeding maximum allowed value
+ */
+static void
+normalize_nsec(int64_t *psecs, int *pnsec)
+{
+	assert(psecs != NULL);
+	assert(pnsec != NULL);
+	int64_t secs = *psecs;
+	int nsec = *pnsec;
+
+	if (nsec < 0 || nsec >= NANOS_PER_SEC) {
+		secs += nsec / NANOS_PER_SEC;
+		nsec %= NANOS_PER_SEC;
+	}
+	*psecs = secs;
+	*pnsec = nsec;
+}
+
+static inline int64_t
+utc_secs(int64_t epoch, int tzoffset)
+{
+	return epoch - tzoffset * 60;
+}
+
+/** minimum supported date - -5879610-06-22 */
+#define MIN_DATE_YEAR -5879610
+#define MIN_DATE_MONTH 6
+#define MIN_DATE_DAY 22
+
+/** maximum supported date - 5879611-07-11 */
+#define MAX_DATE_YEAR 5879611
+#define MAX_DATE_MONTH 7
+#define MAX_DATE_DAY 11
+/**
+ * In the Julian calendar, the average year length is
+ * 365 1/4 days = 365.25 days. This gives an error of
+ * about 1 day in 128 years.
+ */
+#define AVERAGE_DAYS_YEAR 365.25
+#define AVERAGE_DAYS_MONTH (AVERAGE_DAYS_YEAR / 12)
+#define AVERAGE_WEEK_YEAR  (AVERAGE_DAYS_YEAR / 7)
+
+int
+datetime_increment_by(struct datetime *self, int direction,
```

Comment:
These were 3 good examples! Thanks!
- 1st and 3rd examples revealed bugs with internal 32-bit arithmetics overflow. I had to change field types to be able to handle whole supported range of years;
- 2nd is sort of design limitations (of current `c-dt` API). We could only estimate what would be result of operation before `c-dt` function `dt_add_months`/`dt_add_years` called. And this estimation might miss true corner case for operations jumping other boundaries. This could be carefully handled only inside of `c-dt` (similarly as we have done for step.3 in https://github.com/tarantool/c-dt/pull/1/files#diff-89d76e3367f4a376b6c64678a9d08591bb7566bdf05113c9302f1b53921c06fbR154-R189 ) but this case is very, very rare and impractical. So I'd postpone it till next iteration. Not in the nearest RC1;

Reply:
Let me start from test I've developed to check we correctly handle huge ranges here:
```diff
--- a/test/app-tap/datetime.test.lua
+++ b/test/app-tap/datetime.test.lua
@@ -871,7 +871,7 @@ test:test("Time interval operations - different adjustments", function(test)
 end)
 
 test:test("Time intervals creation - range checks", function(test)
-    test:plan(23)
+    test:plan(36)
 
     local inew = date.interval.new
 
@@ -919,6 +919,41 @@ test:test("Time intervals creation - range checks", function(test)
         assert_raises(test, err_msg, function() return inew(attribs) end)
     end
 
+    -- handle tricky cases of huge ranges
+    local huge_year = 3e6
+    local neg_year = date.new{year = -huge_year}
+    test:is(tostring(neg_year), '-3000000-01-01T00:00:00Z', '-3000000')
+    local double_delta = date.interval.new{year = 2 * huge_year}
+    test:is(tostring(double_delta), '+6000000 years', '+6000000 years')
+    local res = neg_year + double_delta
+    test:is(tostring(res), '3000000-01-01T00:00:00Z', '3000000Z')
+    local huge_month = huge_year * 12
+    double_delta = date.interval.new{month = 2 * huge_month}
+    test:is(tostring(double_delta), '+72000000 months', '+72000000 months')
+    res = neg_year + double_delta
+    test:is(tostring(res), '3000000-01-01T00:00:00Z', '3000000Z')
+    local huge_days = huge_year * AVERAGE_DAYS_YEAR
+    double_delta = date.interval.new{day = 2 * huge_days}
+    test:is(tostring(double_delta), '+2191500000 days', '+2191500000 days')
+    res = neg_year + double_delta
+    test:is(tostring(res), '3000123-03-17T00:00:00Z', '3000123-03-17Z')
+    local huge_hours = huge_days * 24
+    double_delta = date.interval.new{hour = 2 * huge_hours}
+    test:is(tostring(double_delta), '+52596000000 hours', '+52596000000 hours')
+    res = neg_year + double_delta
+    test:is(tostring(res), '3000123-03-17T00:00:00Z', '3000123-03-17Z')
...
```

Debugging showed that even for ranges in `day` we might receive values which are not fitting into `int32_t` data type. Furthermore, this problem is getting worse for `hour`, and `min`. There is no such problem for `sec` because it was already 64-bit `double`. Thus we have changed their type and reordered structure layout (to have naturally aligned subfields):

```diff
diff --git a/src/lib/core/datetime.h b/src/lib/core/datetime.h
index 66903190e..dfa913cd6 100644
--- a/src/lib/core/datetime.h
+++ b/src/lib/core/datetime.h
@@ -81,35 +81,34 @@ struct datetime {
 
 /**
  * To be able to perform arithmetics on time intervals and receive
- * deterministic results, we have to keep months and years separately
- * from seconds.
- * Weeks, days, hours and minutes, all could be _precisely_ converted
- * to seconds, but it's not the case for months (which might be 28, 29,
- * 30, or 31 days long), or years (which could be leap year or not).
- * Approach used here - to add/subtract months or years intervals only
- * at the moment when we have particular date we operate on.
- * Determinism of results is achieved due to the order we apply
- * operations (from larger to smaller quantities):
- * - years, then months, then weeks, days, hours, minutes,
- *   seconds, and nanoseconds.
+ * deterministic results, we keep each component (i.e. years, months, weeks,
+ * days, etc) separately from seconds.
+ * We add/subtract interval components separately, and rebase upon resultant
+ * date only at the moment when we apply them to the datetime object, at this
+ * time all leap seconds/leap year logic taken into consideration.
+ * Datetime supports range of values -5879610..5879611, thus interval should be
+ * able to support positive and negative increments covering full distance from
+ * minimum to maximum i.e. 11759221. Such years, months, and weeks values might
+ * be handled by 32-bit integer, but should be extended to 64-bit once we need
+ * to handle corresponding days, hours and seconds values.
  */
 struct datetime_interval {
 	/** Duration in seconds. */
 	double sec;
-	/** Fraction part of duration in seconds. */
-	int nsec;
 	/** Number of minutes, if specified. */
-	int min;
+	double min;
 	/** Number of hours, if specified. */
-	int hour;
+	double hour;
 	/** Number of days, if specified. */
-	int day;
+	double day;
 	/** Number of weeks, if specified. */
-	int week;
+	int32_t week;
 	/** Number of months, if specified. */
-	int month;
+	int32_t month;
 	/** Number of years, if specified. */
-	int year;
+	int32_t year;
+	/** Fraction part of duration in seconds. */
+	int32_t nsec;
 	/** Adjustment mode for day in month operations, @sa dt_adjust_t */
 	dt_adjust_t adjust;
 };
```

As a side effect of those data structure changes we had to change multiple locations here and there (in Lua at the initialization time, in stringization using different output format, in the interval increment function), but patch is too massive to show fully, so it will be collapsed...

<details>
  <summary>Click to expand!</summary>
  
diff --git a/src/lib/core/datetime.c b/src/lib/core/datetime.c
index 466e58110..0b0872579 100644
--- a/src/lib/core/datetime.c
+++ b/src/lib/core/datetime.c
@@ -201,28 +201,6 @@ denormalize_interval_nsec(int64_t *psec, int *pnsec)
 	*pnsec = nsec;
 }
 
-/**
- * SNPRINT-like function to generate textual representation of a sec.nsec
- * given buffer and size of buffer
- */
-static size_t
-seconds_str(char *buf, ssize_t len, bool need_sign, int64_t sec, int nsec)
-{
-	denormalize_interval_nsec(&sec, &nsec);
-	size_t sz = 0;
-	bool is_neg = sec < 0 || (sec == 0 && nsec < 0);
-
-	SNPRINT(sz, snprintf, buf, len, is_neg ? "-" : (need_sign ? "+" : ""));
-	sec = labs(sec);
-	if (nsec != 0) {
-		SNPRINT(sz, snprintf, buf, len, "%" PRId64 ".%09ld seconds",
-			sec, labs(nsec));
-	} else {
-		SNPRINT(sz, snprintf, buf, len, "%" PRId64 " seconds", sec);
-	}
-	return sz;
-}
-
 #define SPACE() \
 	do { \
 		if (sz > 0) { \
@@ -233,6 +211,10 @@ seconds_str(char *buf, ssize_t len, bool need_sign, int64_t sec, int nsec)
 size_t
 interval_to_string(const struct datetime_interval *ival, char *buf, ssize_t len)
 {
+	static const char *const long_signed_fmt[] = {
+		"%" PRId64,	/* false */
+		"%+" PRId64,	/* true */
+	};
 	static const char *const signed_fmt[] = {
 		"%d",	/* false */
 		"%+d",	/* true */
@@ -259,24 +241,27 @@ interval_to_string(const struct datetime_interval *ival, char *buf, ssize_t len)
 		SNPRINT(sz, snprintf, buf, len, " weeks");
 		need_sign = false;
 	}
-	if (ival->day != 0) {
+	int64_t days = (int64_t)ival->day;
+	if (days != 0) {
 		SPACE();
-		SNPRINT(sz, snprintf, buf, len, signed_fmt[need_sign],
-			ival->day);
+		SNPRINT(sz, snprintf, buf, len, long_signed_fmt[need_sign],
+			days);
 		SNPRINT(sz, snprintf, buf, len, " days");
 		need_sign = false;
 	}
+	int64_t hours = (int64_t)ival->hour;
 	if (ival->hour != 0) {
 		SPACE();
-		SNPRINT(sz, snprintf, buf, len, signed_fmt[need_sign],
-			ival->hour);
+		SNPRINT(sz, snprintf, buf, len, long_signed_fmt[need_sign],
+			hours);
 		SNPRINT(sz, snprintf, buf, len, " hours");
 		need_sign = false;
 	}
-	if (ival->min != 0) {
+	int64_t minutes = (int64_t)ival->min;
+	if (minutes != 0) {
 		SPACE();
-		SNPRINT(sz, snprintf, buf, len, signed_fmt[need_sign],
-			ival->min);
+		SNPRINT(sz, snprintf, buf, len, long_signed_fmt[need_sign],
+			minutes);
 		SNPRINT(sz, snprintf, buf, len, " minutes");
 		need_sign = false;
 	}
@@ -289,7 +274,17 @@ interval_to_string(const struct datetime_interval *ival, char *buf, ssize_t len)
 		return sizeof(zero_secs) - 1;
 	}
 	SPACE();
-	SNPRINT(sz, seconds_str, buf, len, need_sign, secs, nsec);
+	denormalize_interval_nsec(&secs, &nsec);
+	bool is_neg = secs < 0 || (secs == 0 && nsec < 0);
+
+	SNPRINT(sz, snprintf, buf, len, is_neg ? "-" : (need_sign ? "+" : ""));
+	secs = labs(secs);
+	if (nsec != 0) {
+		SNPRINT(sz, snprintf, buf, len, "%" PRId64 ".%09ld seconds",
+			secs, labs(nsec));
+	} else {
+		SNPRINT(sz, snprintf, buf, len, "%" PRId64 " seconds", secs);
+	}
 	return sz;
 }
 
@@ -359,22 +354,19 @@ datetime_increment_by(struct datetime *self, int direction,
 	int nsec = self->nsec;
 	int offset = self->tzindex;
 
-	bool is_ym_updated = false;
-	int years = ival->year;
-	int months = ival->month;
-	int weeks = ival->week;
-	int days = ival->day;
-	int hours = ival->hour;
-	int minutes = ival->min;
-	int seconds = ival->sec;
+	bool is_ymd_updated = false;
+	int64_t years = ival->year;
+	int64_t months = ival->month;
+	int64_t weeks = ival->week;
+	int64_t days = ival->day;
+	int64_t hours = ival->hour;
+	int64_t minutes = ival->min;
+	int64_t seconds = ival->sec;
 	int nanoseconds = ival->nsec;
 	dt_adjust_t adjust = ival->adjust;
 	int rc = 0;
 
 	if (years != 0) {
-		rc = verify_range(years, MIN_DATE_YEAR, MAX_DATE_YEAR);
-		if (rc != 0)
-			return rc;
 		rc = verify_dt(dt + direction * years * AVERAGE_DAYS_YEAR);
 		if (rc != 0)
 			return rc;
@@ -382,7 +374,7 @@ datetime_increment_by(struct datetime *self, int direction,
 		 * mode so use tnt_dt_add_months() as a work-around
 		 */
 		dt = dt_add_months(dt, direction * years * 12, adjust);
-		is_ym_updated = true;
+		is_ymd_updated = true;
 	}
 	if (months != 0) {
 		rc = verify_dt(dt + direction * months * AVERAGE_DAYS_MONTH);
@@ -390,7 +382,7 @@ datetime_increment_by(struct datetime *self, int direction,
 			return rc;
 
 		dt = dt_add_months(dt, direction * months, adjust);
-		is_ym_updated = true;
+		is_ymd_updated = true;
 	}
 	if (weeks != 0) {
 		rc = verify_dt(dt + direction * weeks * 7);
@@ -398,7 +390,7 @@ datetime_increment_by(struct datetime *self, int direction,
 			return rc;
 
 		dt += direction * weeks * 7;
-		is_ym_updated = true;
+		is_ymd_updated = true;
 	}
 	if (days != 0) {
 		rc = verify_dt(dt + direction * days);
@@ -406,19 +398,37 @@ datetime_increment_by(struct datetime *self, int direction,
 			return rc;
 
 		dt += direction * days;
-		is_ym_updated = true;
+		is_ymd_updated = true;
 	}
 
-	if (is_ym_updated)
+	if (is_ymd_updated)
 		secs = dt * SECS_PER_DAY - SECS_EPOCH_1970_OFFSET +
 			secs % SECS_PER_DAY;
 
-	if (hours != 0)
+	if (hours != 0) {
+		rc = verify_range(secs + direction * hours * 3600,
+				  MIN_EPOCH_SECS_VALUE, MAX_EPOCH_SECS_VALUE);
+		if (rc != 0)
+			return rc;
+
 		secs += direction * hours * 3600;
-	if (minutes != 0)
+	}
+	if (minutes != 0) {
+		rc = verify_range(secs + direction * minutes * 60,
+				  MIN_EPOCH_SECS_VALUE, MAX_EPOCH_SECS_VALUE);
+		if (rc != 0)
+			return rc;
+
 		secs += direction * minutes * 60;
-	if (seconds != 0)
+	}
+	if (seconds != 0) {
+		rc = verify_range(secs + direction * seconds,
+				  MIN_EPOCH_SECS_VALUE, MAX_EPOCH_SECS_VALUE);
+		if (rc != 0)
+			return rc;
+
 		secs += direction * seconds;
+	}
 	if (nanoseconds != 0)
 		nsec += direction * nanoseconds;
 
diff --git a/src/lua/datetime.lua b/src/lua/datetime.lua
index 18ffe17f1..9c059a434 100644
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -300,7 +300,7 @@ local adjust_xlat = {
 }
 
 local function interval_init(year, month, week, day, hour, min, sec, nsec, adjust)
-    return ffi.new(interval_t, sec, nsec, min, hour, day, week, month, year, adjust)
+    return ffi.new(interval_t, sec, min, hour, day, week, month, year, nsec, adjust)
 end
 
 local function interval_new_copy(obj)
diff --git a/src/lua/utils.c b/src/lua/utils.c
index 31de56aad..f5849badc 100644
--- a/src/lua/utils.c
+++ b/src/lua/utils.c
@@ -752,14 +752,14 @@ tarantool_lua_utils_init(struct lua_State *L)
 
 	rc = luaL_cdef(L, "struct datetime_interval {"
 			  "double sec;"
-			  "int nsec;"
-			  "int min;"
-			  "int hour;"
-			  "int day;"
-			  "int week;"
-			  "int month;"
-			  "int year;"
-			  "int adjust;"
+			  "double min;"
+			  "double hour;"
+			  "double day;"
+			  "int32_t week;"
+			  "int32_t month;"
+			  "int32_t year;"
+			  "int32_t nsec;"
+			  "dt_adjust_t adjust;"
 			  "};");
 	assert(rc == 0);
 	(void)rc;
</details>